# Simulations

## ath

Input: selection of *Arabidopsis* CDS
Spontaneous mutation rate: *Arabidopsis* (7 x 10-9 substitution per site per generation of 1 year)
Indels: 0.0857 deletions relative to substitution rates and 0.0429 insertions relative to substitution rates (*Arabidopsis*)
Alignment program: Alignment using tfasty

## ath_x_10

Input: selection of *Arabidopsis* CDS
Spontaneous mutation rate: *Arabidopsis* (7 x 10-9 substitution per site per generation of 1 year) * 10
Indels: 0.0857 deletions relative to substitution rates and 0.0429 insertions relative to substitution rates (*Arabidopsis*)
Alignment program: Alignment using tfasty

## ath_div_10

Input: selection of *Arabidopsis* CDS
Spontaneous mutation rate: *Arabidopsis* (7 x 10-9 substitution per site per generation of 1 year) / 10
Indels: 0.0857 deletions relative to substitution rates and 0.0429 insertions relative to substitution rates (*Arabidopsis*)
Alignment program: Alignment using tfasty

## ptr

Input: selection of *Arabidopsis* CDS
Spontaneous mutation rate: *Populus* (1.33 x 10-10 substitution per site per 1 year)
Indels: 0.0857 deletions relative to substitution rates and 0.0429 insertions relative to substitution rates (*Arabidopsis*)
Alignment program: Alignment using tfasty

## ath_x_2

Input: selection of *Arabidopsis* CDS
Spontaneous mutation rate: *Arabidopsis* (7 x 10-9 substitution per site per generation of 1 year) X 2
        as if both genes evolve as pseudogenes
Indels: 0.0857 deletions relative to substitution rates and 0.0429 insertions relative to substitution rates (*Arabidopsis*)
Alignment program: Alignment using tfasty

## ptr_x_2

Input: selection of *Arabidopsis* CDS
Spontaneous mutation rate: *Populus* (1.33 x 10-10 substitution per site per 1 year) X 2
        as if both genes evolve as pseudogenes
Indels: 0.0857 deletions relative to substitution rates and 0.0429 insertions relative to substitution rates (*Arabidopsis*)
Alignment program: Alignment using tfasty